File: examples/typescript/customer/src/command-validation-error.ts
/**
 * Error class for command validation failures.
 * Bridges domain validation to gRPC status codes without coupling logic to gRPC.
 */
export enum StatusCode {
  INVALID_ARGUMENT = 'INVALID_ARGUMENT',
  FAILED_PRECONDITION = 'FAILED_PRECONDITION',
}

export class CommandValidationError extends Error {
  constructor(
    message: string,
    public readonly statusCode: StatusCode
  ) {
    super(message);
    this.name = 'CommandValidationError';
  }

  static invalidArgument(message: string): CommandValidationError {
    return new CommandValidationError(message, StatusCode.INVALID_ARGUMENT);
  }

  static failedPrecondition(message: string): CommandValidationError {
    return new CommandValidationError(message, StatusCode.FAILED_PRECONDITION);
  }
}

File: examples/ruby/inventory/lib/command_validation_error.rb
# frozen_string_literal: true

# Error class for command validation failures.
# Bridges domain validation to gRPC status codes without coupling logic to gRPC.
class CommandValidationError < StandardError
  attr_reader :status_code

  INVALID_ARGUMENT = :invalid_argument
  FAILED_PRECONDITION = :failed_precondition

  def initialize(message, status_code)
    super(message)
    @status_code = status_code
  end

  def self.invalid_argument(message)
    new(message, INVALID_ARGUMENT)
  end

  def self.failed_precondition(message)
    new(message, FAILED_PRECONDITION)
  end
end
